### ISSUE TITLE
Blank nodes are not deskolemized

### ISSUE THREAD

**Issue by smessie** [comunica/comunica#1057]
#### Issue type:

- :bug: Bug

____
#### Description:

It looks like blank nodes are not deskolemized after they were skolemized. While debugging why the INSERT-DELETE-WHERE was not working, we saw that (also) for a simple SELECT the blank nodes are not deskolemized. This is most likely also the reason why the SPARQL patch is not updating anything; the values of the nodes are still prefixed with `bc_0`.

A small example to reproduce the bug might clear things up about what is the issue.

```javascript
const rdflib = require('rdflib')
const QueryEngine = require('@comunica/query-sparql').QueryEngine;
const N3 = require('n3');

let _store = rdflib.graph()
let body = `
<#78012f06-5e20-4a58-9339-74ee0288f608> a <http://www.w3.org/ns/hydra/core#Class>;
    <http://www.w3.org/ns/hydra/core#method> "POST".
<#34b182dd-7d2f-4eaf-b604-5f4f4470cec3> a <http://rdf.danielbeeke.nl/form/form-dev.ttl#Form>;
    <http://www.w3.org/ns/hydra/core#endpoint> <https://httpbin.org/post>;
    <http://www.w3.org/ns/hydra/core#supportedClass> <#78012f06-5e20-4a58-9339-74ee0288f608>;
    <http://rdf.danielbeeke.nl/form/form-dev.ttl#binding> <https://example.org/#BlankNodesForm>.
<#78926887-5eb7-4427-8512-0d3b828c1404> a <http://rdf.danielbeeke.nl/form/form-dev.ttl#Field>;
    <http://rdf.danielbeeke.nl/form/form-dev.ttl#binding> <https://example.org/#Select>;
    <http://rdf.danielbeeke.nl/form/form-dev.ttl#widget> "dropdown";
    <http://rdf.danielbeeke.nl/form/form-dev.ttl#label> "Select";
    <http://rdf.danielbeeke.nl/form/form-dev.ttl#required> false;
    <http://rdf.danielbeeke.nl/form/form-dev.ttl#multiple> false;
    <http://rdf.danielbeeke.nl/form/form-dev.ttl#option> _:b17145_b17132_b17119_e_n101250.
_:b17145_b17132_b17119_e_n101250 <http://www.w3.org/1999/02/22-rdf-syntax-ns#rest> _:b17145_b17132_b17119_e_n100250.
_:b17145_b17132_b17119_e_n101250 <http://www.w3.org/1999/02/22-rdf-syntax-ns#first> <#c20029de-6509-49e0-8333-7e80e17157da>.
_:b17145_b17132_b17119_e_n100250 <http://www.w3.org/1999/02/22-rdf-syntax-ns#rest> <http://www.w3.org/1999/02/22-rdf-syntax-ns#nil>;
    <http://www.w3.org/1999/02/22-rdf-syntax-ns#first> <#bc4d1a83-81b0-48ff-96b8-1eb208c13869>.
<#c20029de-6509-49e0-8333-7e80e17157da> a <http://rdf.danielbeeke.nl/form/form-dev.ttl#Option>;
    <http://rdf.danielbeeke.nl/form/form-dev.ttl#label> "One";
    <http://rdf.danielbeeke.nl/form/form-dev.ttl#value> <https://example.org/#One>.
<#bc4d1a83-81b0-48ff-96b8-1eb208c13869> a <http://rdf.danielbeeke.nl/form/form-dev.ttl#Option>;
    <http://rdf.danielbeeke.nl/form/form-dev.ttl#label> "Two";
    <http://rdf.danielbeeke.nl/form/form-dev.ttl#value> <https://example.org/#Two>.
`
let mimeType = 'text/turtle'
let uri = 'http://local'

rdflib.parse(body, _store, uri, mimeType)


async function test() {
    const myEngine = new QueryEngine();
    // const store = new N3.Store();
    let store = new N3.Store()
    for (const e of _store.match()) {
        console.log(e)
        store.addQuad(e.subject, e.predicate, e.object)
    }
    console.log(store.getQuads())

    await myEngine.queryVoid(`
    DELETE { ?var1 <http://www.w3.org/1999/02/22-rdf-syntax-ns#rest> ?var2 . }
    INSERT { ?var2 <http://www.w3.org/1999/02/22-rdf-syntax-ns#rest> ?var1 .
     } WHERE { ?var1 <http://www.w3.org/1999/02/22-rdf-syntax-ns#rest> ?var2 .}
    `, {
        sources: [store],
    });
    const {data} = await myEngine.resultToString(await myEngine.query(`SELECT * WHERE {?o <http://www.w3.org/1999/02/22-rdf-syntax-ns#rest> ?s}`, {
            sources: [store],
        }),
        'application/sparql-results+json');
    console.log(`Output data is`)
    data.pipe(process.stdout)
}
test()
```

The output is:
<details>
<summary>Full output of `store.getQuads()`</summary>

```
Statement {
  subject: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'http://local#78012f06-5e20-4a58-9339-74ee0288f608'
  },
  predicate: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'http://www.w3.org/1999/02/22-rdf-syntax-ns#type'
  },
  object: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'http://www.w3.org/ns/hydra/core#Class'
  },
  graph: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'http://local'
  }
}
Statement {
  subject: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'http://local#78012f06-5e20-4a58-9339-74ee0288f608'
  },
  predicate: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'http://www.w3.org/ns/hydra/core#method'
  },
  object: Literal {
    termType: 'Literal',
    classOrder: 1,
    value: 'POST',
    datatype: NamedNode {
      termType: 'NamedNode',
      classOrder: 5,
      value: 'http://www.w3.org/2001/XMLSchema#string'
    },
    isVar: 0,
    language: ''
  },
  graph: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'http://local'
  }
}
Statement {
  subject: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'http://local#34b182dd-7d2f-4eaf-b604-5f4f4470cec3'
  },
  predicate: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'http://www.w3.org/1999/02/22-rdf-syntax-ns#type'
  },
  object: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'http://rdf.danielbeeke.nl/form/form-dev.ttl#Form'
  },
  graph: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'http://local'
  }
}
Statement {
  subject: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'http://local#34b182dd-7d2f-4eaf-b604-5f4f4470cec3'
  },
  predicate: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'http://www.w3.org/ns/hydra/core#endpoint'
  },
  object: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'https://httpbin.org/post'
  },
  graph: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'http://local'
  }
}
Statement {
  subject: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'http://local#34b182dd-7d2f-4eaf-b604-5f4f4470cec3'
  },
  predicate: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'http://www.w3.org/ns/hydra/core#supportedClass'
  },
  object: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'http://local#78012f06-5e20-4a58-9339-74ee0288f608'
  },
  graph: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'http://local'
  }
}
Statement {
  subject: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'http://local#34b182dd-7d2f-4eaf-b604-5f4f4470cec3'
  },
  predicate: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'http://rdf.danielbeeke.nl/form/form-dev.ttl#binding'
  },
  object: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'https://example.org/#BlankNodesForm'
  },
  graph: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'http://local'
  }
}
Statement {
  subject: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'http://local#78926887-5eb7-4427-8512-0d3b828c1404'
  },
  predicate: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'http://www.w3.org/1999/02/22-rdf-syntax-ns#type'
  },
  object: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'http://rdf.danielbeeke.nl/form/form-dev.ttl#Field'
  },
  graph: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'http://local'
  }
}
Statement {
  subject: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'http://local#78926887-5eb7-4427-8512-0d3b828c1404'
  },
  predicate: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'http://rdf.danielbeeke.nl/form/form-dev.ttl#binding'
  },
  object: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'https://example.org/#Select'
  },
  graph: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'http://local'
  }
}
Statement {
  subject: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'http://local#78926887-5eb7-4427-8512-0d3b828c1404'
  },
  predicate: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'http://rdf.danielbeeke.nl/form/form-dev.ttl#widget'
  },
  object: Literal {
    termType: 'Literal',
    classOrder: 1,
    value: 'dropdown',
    datatype: NamedNode {
      termType: 'NamedNode',
      classOrder: 5,
      value: 'http://www.w3.org/2001/XMLSchema#string'
    },
    isVar: 0,
    language: ''
  },
  graph: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'http://local'
  }
}
Statement {
  subject: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'http://local#78926887-5eb7-4427-8512-0d3b828c1404'
  },
  predicate: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'http://rdf.danielbeeke.nl/form/form-dev.ttl#label'
  },
  object: Literal {
    termType: 'Literal',
    classOrder: 1,
    value: 'Select',
    datatype: NamedNode {
      termType: 'NamedNode',
      classOrder: 5,
      value: 'http://www.w3.org/2001/XMLSchema#string'
    },
    isVar: 0,
    language: ''
  },
  graph: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'http://local'
  }
}
Statement {
  subject: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'http://local#78926887-5eb7-4427-8512-0d3b828c1404'
  },
  predicate: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'http://rdf.danielbeeke.nl/form/form-dev.ttl#required'
  },
  object: Literal {
    termType: 'Literal',
    classOrder: 1,
    value: '0',
    datatype: NamedNode {
      termType: 'NamedNode',
      classOrder: 5,
      value: 'http://www.w3.org/2001/XMLSchema#boolean'
    },
    isVar: 0,
    language: ''
  },
  graph: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'http://local'
  }
}
Statement {
  subject: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'http://local#78926887-5eb7-4427-8512-0d3b828c1404'
  },
  predicate: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'http://rdf.danielbeeke.nl/form/form-dev.ttl#multiple'
  },
  object: Literal {
    termType: 'Literal',
    classOrder: 1,
    value: '0',
    datatype: NamedNode {
      termType: 'NamedNode',
      classOrder: 5,
      value: 'http://www.w3.org/2001/XMLSchema#boolean'
    },
    isVar: 0,
    language: ''
  },
  graph: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'http://local'
  }
}
Statement {
  subject: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'http://local#78926887-5eb7-4427-8512-0d3b828c1404'
  },
  predicate: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'http://rdf.danielbeeke.nl/form/form-dev.ttl#option'
  },
  object: BlankNode {
    termType: 'BlankNode',
    classOrder: 6,
    value: 'b17145_b17132_b17119_e_n101250',
    isBlank: 1,
    isVar: 1
  },
  graph: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'http://local'
  }
}
Statement {
  subject: BlankNode {
    termType: 'BlankNode',
    classOrder: 6,
    value: 'b17145_b17132_b17119_e_n101250',
    isBlank: 1,
    isVar: 1
  },
  predicate: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'http://www.w3.org/1999/02/22-rdf-syntax-ns#rest'
  },
  object: BlankNode {
    termType: 'BlankNode',
    classOrder: 6,
    value: 'b17145_b17132_b17119_e_n100250',
    isBlank: 1,
    isVar: 1
  },
  graph: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'http://local'
  }
}
Statement {
  subject: BlankNode {
    termType: 'BlankNode',
    classOrder: 6,
    value: 'b17145_b17132_b17119_e_n101250',
    isBlank: 1,
    isVar: 1
  },
  predicate: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'http://www.w3.org/1999/02/22-rdf-syntax-ns#first'
  },
  object: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'http://local#c20029de-6509-49e0-8333-7e80e17157da'
  },
  graph: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'http://local'
  }
}
Statement {
  subject: BlankNode {
    termType: 'BlankNode',
    classOrder: 6,
    value: 'b17145_b17132_b17119_e_n100250',
    isBlank: 1,
    isVar: 1
  },
  predicate: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'http://www.w3.org/1999/02/22-rdf-syntax-ns#rest'
  },
  object: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'http://www.w3.org/1999/02/22-rdf-syntax-ns#nil'
  },
  graph: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'http://local'
  }
}
Statement {
  subject: BlankNode {
    termType: 'BlankNode',
    classOrder: 6,
    value: 'b17145_b17132_b17119_e_n100250',
    isBlank: 1,
    isVar: 1
  },
  predicate: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'http://www.w3.org/1999/02/22-rdf-syntax-ns#first'
  },
  object: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'http://local#bc4d1a83-81b0-48ff-96b8-1eb208c13869'
  },
  graph: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'http://local'
  }
}
Statement {
  subject: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'http://local#c20029de-6509-49e0-8333-7e80e17157da'
  },
  predicate: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'http://www.w3.org/1999/02/22-rdf-syntax-ns#type'
  },
  object: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'http://rdf.danielbeeke.nl/form/form-dev.ttl#Option'
  },
  graph: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'http://local'
  }
}
Statement {
  subject: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'http://local#c20029de-6509-49e0-8333-7e80e17157da'
  },
  predicate: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'http://rdf.danielbeeke.nl/form/form-dev.ttl#label'
  },
  object: Literal {
    termType: 'Literal',
    classOrder: 1,
    value: 'One',
    datatype: NamedNode {
      termType: 'NamedNode',
      classOrder: 5,
      value: 'http://www.w3.org/2001/XMLSchema#string'
    },
    isVar: 0,
    language: ''
  },
  graph: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'http://local'
  }
}
Statement {
  subject: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'http://local#c20029de-6509-49e0-8333-7e80e17157da'
  },
  predicate: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'http://rdf.danielbeeke.nl/form/form-dev.ttl#value'
  },
  object: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'https://example.org/#One'
  },
  graph: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'http://local'
  }
}
Statement {
  subject: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'http://local#bc4d1a83-81b0-48ff-96b8-1eb208c13869'
  },
  predicate: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'http://www.w3.org/1999/02/22-rdf-syntax-ns#type'
  },
  object: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'http://rdf.danielbeeke.nl/form/form-dev.ttl#Option'
  },
  graph: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'http://local'
  }
}
Statement {
  subject: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'http://local#bc4d1a83-81b0-48ff-96b8-1eb208c13869'
  },
  predicate: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'http://rdf.danielbeeke.nl/form/form-dev.ttl#label'
  },
  object: Literal {
    termType: 'Literal',
    classOrder: 1,
    value: 'Two',
    datatype: NamedNode {
      termType: 'NamedNode',
      classOrder: 5,
      value: 'http://www.w3.org/2001/XMLSchema#string'
    },
    isVar: 0,
    language: ''
  },
  graph: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'http://local'
  }
}
Statement {
  subject: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'http://local#bc4d1a83-81b0-48ff-96b8-1eb208c13869'
  },
  predicate: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'http://rdf.danielbeeke.nl/form/form-dev.ttl#value'
  },
  object: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'https://example.org/#Two'
  },
  graph: NamedNode {
    termType: 'NamedNode',
    classOrder: 5,
    value: 'http://local'
  }
}
[
  Quad {
    id: '',
    _subject: NamedNode {
      id: 'http://local#78012f06-5e20-4a58-9339-74ee0288f608'
    },
    _predicate: NamedNode { id: 'http://www.w3.org/1999/02/22-rdf-syntax-ns#type' },
    _object: NamedNode { id: 'http://www.w3.org/ns/hydra/core#Class' },
    _graph: DefaultGraph { id: '' }
  },
  Quad {
    id: '',
    _subject: NamedNode {
      id: 'http://local#78012f06-5e20-4a58-9339-74ee0288f608'
    },
    _predicate: NamedNode { id: 'http://www.w3.org/ns/hydra/core#method' },
    _object: Literal { id: '"POST"' },
    _graph: DefaultGraph { id: '' }
  },
  Quad {
    id: '',
    _subject: NamedNode {
      id: 'http://local#34b182dd-7d2f-4eaf-b604-5f4f4470cec3'
    },
    _predicate: NamedNode { id: 'http://www.w3.org/1999/02/22-rdf-syntax-ns#type' },
    _object: NamedNode {
      id: 'http://rdf.danielbeeke.nl/form/form-dev.ttl#Form'
    },
    _graph: DefaultGraph { id: '' }
  },
  Quad {
    id: '',
    _subject: NamedNode {
      id: 'http://local#34b182dd-7d2f-4eaf-b604-5f4f4470cec3'
    },
    _predicate: NamedNode { id: 'http://www.w3.org/ns/hydra/core#endpoint' },
    _object: NamedNode { id: 'https://httpbin.org/post' },
    _graph: DefaultGraph { id: '' }
  },
  Quad {
    id: '',
    _subject: NamedNode {
      id: 'http://local#34b182dd-7d2f-4eaf-b604-5f4f4470cec3'
    },
    _predicate: NamedNode { id: 'http://www.w3.org/ns/hydra/core#supportedClass' },
    _object: NamedNode {
      id: 'http://local#78012f06-5e20-4a58-9339-74ee0288f608'
    },
    _graph: DefaultGraph { id: '' }
  },
  Quad {
    id: '',
    _subject: NamedNode {
      id: 'http://local#34b182dd-7d2f-4eaf-b604-5f4f4470cec3'
    },
    _predicate: NamedNode {
      id: 'http://rdf.danielbeeke.nl/form/form-dev.ttl#binding'
    },
    _object: NamedNode { id: 'https://example.org/#BlankNodesForm' },
    _graph: DefaultGraph { id: '' }
  },
  Quad {
    id: '',
    _subject: NamedNode {
      id: 'http://local#78926887-5eb7-4427-8512-0d3b828c1404'
    },
    _predicate: NamedNode { id: 'http://www.w3.org/1999/02/22-rdf-syntax-ns#type' },
    _object: NamedNode {
      id: 'http://rdf.danielbeeke.nl/form/form-dev.ttl#Field'
    },
    _graph: DefaultGraph { id: '' }
  },
  Quad {
    id: '',
    _subject: NamedNode {
      id: 'http://local#78926887-5eb7-4427-8512-0d3b828c1404'
    },
    _predicate: NamedNode {
      id: 'http://rdf.danielbeeke.nl/form/form-dev.ttl#binding'
    },
    _object: NamedNode { id: 'https://example.org/#Select' },
    _graph: DefaultGraph { id: '' }
  },
  Quad {
    id: '',
    _subject: NamedNode {
      id: 'http://local#78926887-5eb7-4427-8512-0d3b828c1404'
    },
    _predicate: NamedNode {
      id: 'http://rdf.danielbeeke.nl/form/form-dev.ttl#widget'
    },
    _object: Literal { id: '"dropdown"' },
    _graph: DefaultGraph { id: '' }
  },
  Quad {
    id: '',
    _subject: NamedNode {
      id: 'http://local#78926887-5eb7-4427-8512-0d3b828c1404'
    },
    _predicate: NamedNode {
      id: 'http://rdf.danielbeeke.nl/form/form-dev.ttl#label'
    },
    _object: Literal { id: '"Select"' },
    _graph: DefaultGraph { id: '' }
  },
  Quad {
    id: '',
    _subject: NamedNode {
      id: 'http://local#78926887-5eb7-4427-8512-0d3b828c1404'
    },
    _predicate: NamedNode {
      id: 'http://rdf.danielbeeke.nl/form/form-dev.ttl#required'
    },
    _object: Literal { id: '"0"^^http://www.w3.org/2001/XMLSchema#boolean' },
    _graph: DefaultGraph { id: '' }
  },
  Quad {
    id: '',
    _subject: NamedNode {
      id: 'http://local#78926887-5eb7-4427-8512-0d3b828c1404'
    },
    _predicate: NamedNode {
      id: 'http://rdf.danielbeeke.nl/form/form-dev.ttl#multiple'
    },
    _object: Literal { id: '"0"^^http://www.w3.org/2001/XMLSchema#boolean' },
    _graph: DefaultGraph { id: '' }
  },
  Quad {
    id: '',
    _subject: NamedNode {
      id: 'http://local#78926887-5eb7-4427-8512-0d3b828c1404'
    },
    _predicate: NamedNode {
      id: 'http://rdf.danielbeeke.nl/form/form-dev.ttl#option'
    },
    _object: BlankNode { id: '_:b17145_b17132_b17119_e_n101250' },
    _graph: DefaultGraph { id: '' }
  },
  Quad {
    id: '',
    _subject: BlankNode { id: '_:b17145_b17132_b17119_e_n101250' },
    _predicate: NamedNode { id: 'http://www.w3.org/1999/02/22-rdf-syntax-ns#rest' },
    _object: BlankNode { id: '_:b17145_b17132_b17119_e_n100250' },
    _graph: DefaultGraph { id: '' }
  },
  Quad {
    id: '',
    _subject: BlankNode { id: '_:b17145_b17132_b17119_e_n101250' },
    _predicate: NamedNode {
      id: 'http://www.w3.org/1999/02/22-rdf-syntax-ns#first'
    },
    _object: NamedNode {
      id: 'http://local#c20029de-6509-49e0-8333-7e80e17157da'
    },
    _graph: DefaultGraph { id: '' }
  },
  Quad {
    id: '',
    _subject: BlankNode { id: '_:b17145_b17132_b17119_e_n100250' },
    _predicate: NamedNode { id: 'http://www.w3.org/1999/02/22-rdf-syntax-ns#rest' },
    _object: NamedNode { id: 'http://www.w3.org/1999/02/22-rdf-syntax-ns#nil' },
    _graph: DefaultGraph { id: '' }
  },
  Quad {
    id: '',
    _subject: BlankNode { id: '_:b17145_b17132_b17119_e_n100250' },
    _predicate: NamedNode {
      id: 'http://www.w3.org/1999/02/22-rdf-syntax-ns#first'
    },
    _object: NamedNode {
      id: 'http://local#bc4d1a83-81b0-48ff-96b8-1eb208c13869'
    },
    _graph: DefaultGraph { id: '' }
  },
  Quad {
    id: '',
    _subject: NamedNode {
      id: 'http://local#c20029de-6509-49e0-8333-7e80e17157da'
    },
    _predicate: NamedNode { id: 'http://www.w3.org/1999/02/22-rdf-syntax-ns#type' },
    _object: NamedNode {
      id: 'http://rdf.danielbeeke.nl/form/form-dev.ttl#Option'
    },
    _graph: DefaultGraph { id: '' }
  },
  Quad {
    id: '',
    _subject: NamedNode {
      id: 'http://local#c20029de-6509-49e0-8333-7e80e17157da'
    },
    _predicate: NamedNode {
      id: 'http://rdf.danielbeeke.nl/form/form-dev.ttl#label'
    },
    _object: Literal { id: '"One"' },
    _graph: DefaultGraph { id: '' }
  },
  Quad {
    id: '',
    _subject: NamedNode {
      id: 'http://local#c20029de-6509-49e0-8333-7e80e17157da'
    },
    _predicate: NamedNode {
      id: 'http://rdf.danielbeeke.nl/form/form-dev.ttl#value'
    },
    _object: NamedNode { id: 'https://example.org/#One' },
    _graph: DefaultGraph { id: '' }
  },
  Quad {
    id: '',
    _subject: NamedNode {
      id: 'http://local#bc4d1a83-81b0-48ff-96b8-1eb208c13869'
    },
    _predicate: NamedNode { id: 'http://www.w3.org/1999/02/22-rdf-syntax-ns#type' },
    _object: NamedNode {
      id: 'http://rdf.danielbeeke.nl/form/form-dev.ttl#Option'
    },
    _graph: DefaultGraph { id: '' }
  },
  Quad {
    id: '',
    _subject: NamedNode {
      id: 'http://local#bc4d1a83-81b0-48ff-96b8-1eb208c13869'
    },
    _predicate: NamedNode {
      id: 'http://rdf.danielbeeke.nl/form/form-dev.ttl#label'
    },
    _object: Literal { id: '"Two"' },
    _graph: DefaultGraph { id: '' }
  },
  Quad {
    id: '',
    _subject: NamedNode {
      id: 'http://local#bc4d1a83-81b0-48ff-96b8-1eb208c13869'
    },
    _predicate: NamedNode {
      id: 'http://rdf.danielbeeke.nl/form/form-dev.ttl#value'
    },
    _object: NamedNode { id: 'https://example.org/#Two' },
    _graph: DefaultGraph { id: '' }
  }
]
```

</details>

```
Output data is
{"head": {"vars":["o","s"]},
"results": { "bindings": [
{"o":{"value":"bc_0_b17145_b17132_b17119_e_n100250","type":"bnode"},"s":{"value":"bc_0_b17145_b17132_b17119_e_n101250","type":"bnode"}},
{"o":{"value":"http://www.w3.org/1999/02/22-rdf-syntax-ns#nil","type":"uri"},"s":{"value":"bc_0_b17145_b17132_b17119_e_n100250","type":"bnode"}}
]},
"metadata": { "httpRequests": 0 }}
```

Mind the `bc_0_b17145_b17132_b17119_e_n100250` instead of `b17145_b17132_b17119_e_n100250`.

Credits to @Jan-PieterBaert for helping debugging

____
#### Environment:

| software         | version
| ---------------- | -------
| Comunica Engine  | 2.4.1 
| node             | v16.16.0
| npm              | 8.14.0
| yarn             | 1.22.19
| Operating System | darwin (Darwin 21.6.0)

**Comment by rubensworks** [comment]
Deskolemization before updates should have been added in #922.

@jeswr does this issue seem related to the one fixed in #922, or is it a different one?

**Comment by jeswr** [comment]
Should be resolved by #922 - I added an integration test [very similar](https://github.com/jeswr/comunica/blob/5aaad22db39f8e3a3c1f0d7b559511c5ba7687d8/engines/query-sparql/test/QuerySparql-test.ts#L394-L423) to this at the time.

Might be worth adding the above as an integration test - but beyond that it should be safe to close this issue.

**Comment by smessie** [comment]
> TLDR; Since you only have one source try executing the SELECT query with
> 
> ```ts
> {
>         source: store
> }
> ```
> 
> rather than
> 
> ```ts
> {
>         sources: [store],
> }
> ```
> 
> this way it won't pass through the federated components which are adding this in.

Then I get following error:
```
/Users/ieben/WebstormProjects/comunica-sparql-patch-test/node_modules/@comunica/mediator-race/lib/MediatorRace.js:22
                        reject(new Error(`${this.name} mediated over all rejecting actors:\n${errors.map(subError => subError.message).join('\n')}`));
                               ^

Error: urn:comunica:default:rdf-update-quads/mediators#main mediated over all rejecting actors:
urn:comunica:default:rdf-update-quads/actors#rdfjs-store received an invalid rdfjsStore.
Actor urn:comunica:default:rdf-update-quads/actors#hypermedia can only update quads against a single destination URL.
    at /Users/ieben/WebstormProjects/comunica-sparql-patch-test/node_modules/@comunica/mediator-race/lib/MediatorRace.js:22:32
    at processTicksAndRejections (node:internal/process/task_queues:96:5)

```
With following warning in my IDE
```
Argument type {source: N3Store} is not assignable to parameter type " DELETE { ?var1 <http://www.w3.org/1999/02/22-rdf-syntax-ns#rest> ?var2 . } INSERT { ?var2 <http://www.w3.org/1999/02/22-rdf-syntax-ns#rest> ?var1 . } WHERE { ?var1 <http://www.w3.org/1999/02/22-rdf-syntax-ns#rest> ?var2 .} " extends string ? QueryStringContext : QueryAlgebraContext | undefined 
```

---

> I don't think there problem here is to do with the patching query. To me, it appears that the blank nodes are correctly being Skolemised before being returned from the SELECT query - in the federated case the `bc_0` _needs_ to be there to prevent clashes between naming of blank nodes between documents and hence breaking of blank node semantics between documents.

Okay, I do understand your point that you skolemize it to cover the case of multiple sources. However, this `SELECT` example resulted in the fact that we had problems with a `DELETE - INSERT - WHERE` that did not have any effect, so we constructed this SELECT as an easier case to show the problem.
Could it be that the same problem appears in a SPARQL patch? That it tries to delete the skolemized triples and thus cannot find them in the store?
I would want to debug a bit, but I have a hard time setting up the dev environment.

**Comment by jeswr** [comment]
> Then I get following error:

Huh - I was assuming the context preprocessor behaved differently; see #1058 .

> Could it be that the same problem appears in a SPARQL patch? That it tries to delete the skolemized triples and thus cannot find them in the store?

`DELETE - INSERT-WHERE` problems should have been resolved by https://github.com/comunica/comunica/pull/922 - I added an integration test [very similar](https://github.com/jeswr/comunica/blob/5aaad22db39f8e3a3c1f0d7b559511c5ba7687d8/engines/query-sparql/test/QuerySparql-test.ts#L345-L383) to this at the time. If you can give me a more minimal example involving on `DELETE - INSERT-WHERE` query and one or 2 triples that would be great. I'm pushed for time to debug a bigger example rn.

All quad updates whether `DELETE - INSERT - WHERE` or `INSERT DATA` run through [this](https://github.com/comunica/comunica/blob/a9eaf5db72f4cd5364cdb0dda9eb5fdb7996a125/packages/bus-rdf-update-quads/lib/ActorRdfUpdateQuadsDestination.ts#L41-L42) deskolemisation process which will strip the source id if there is one attached; and otherwise use the original blank node identifier due to this line https://github.com/comunica/comunica/blob/a9eaf5db72f4cd5364cdb0dda9eb5fdb7996a125/packages/actor-rdf-resolve-quad-pattern-federated/lib/FederatedQuadSource.ts#L144. I do not see any errors with this logic; please point them out to me if there are any.

**Comment by rubensworks** [comment]
Related to #985
